src/Chapter3.hs

```diff
@@ -1006,6 +1128,99 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+class Fighter a where
+  attack :: a -> Int
+  health  :: a -> Int
+  takeTurn :: a -> Fightable -> Fight
+  receiveDamage :: a -> Int -> a
+
+data Fightable =
+    FightHuman Human
+  | FightBeast Beast
+    deriving Show
+
+data Fight = Fight Fightable Fightable
+  deriving Show
+
+instance Fighter Fightable where
+  attack :: Fightable -> Int
+  attack (FightHuman human) = humanAtk human
+  attack (FightBeast beast) = beastAtk beast
+
+  health :: Fightable -> Int
+  health (FightHuman human) = humanHp human
+  health (FightBeast beast) = beastHp beast
+
+  receiveDamage :: Fightable -> Int -> Fightable
+  receiveDamage (FightHuman human) dmg = FightHuman (receiveDamage human dmg)
+  receiveDamage (FightBeast beast) dmg = FightBeast (receiveDamage beast dmg)
+
```

Comment:
The above implementation of `receiveDamage` could almost be just one line, but I don't know how to reach inside of the `Fightable` given as an argument without pattern matching on a specific `Fightable` constructor. Is it possible?

Reply:
If you would have a polymorphic variable instead of two constructors, you would be able to add `Functor` instance to the `Fightable` data type (we will cover Functors in the next Chapter). Then you would be able to write something like 
```haskell
receiveDamage fighter dmg = fmap (`receiveDamage` dmg) fighter
```

Reply:
To be honest I do not completely follow the idea of having `Fightable` as a sum type instead of using the `Fighter` typeclass 🤔  (the naming seems to be mixed up, as typeclass is usually something action-able, and the data type is a noun 😄 )

Reply:
Thank you for the review! I agree my solution is confusing. I will try to refactor it some time when work has slowed down :)

Reply:
No worries @OxSon ! You did a great job, and I really enjoyed going through your solutions as they are very thoughtful 👌🏼 Typeclasses are not the easy topic so it always requires some time to find the optimal balance in methods in the class and additional functions that work with the types that have these instances ⚖️